### ISSUE TITLE
Maybe use front-matter yaml markup?

### ISSUE THREAD

**Issue by felipesanches** [fonttools/fontbakery#1843]
@graphicore suggested it at https://github.com/googlefonts/fontbakery/issues/1584#issuecomment-385244577

It is not clear exactly what would be the benefit of using it. Perhaps we could take a look at an example of how it would look like? Could you provide an example, Lasse?

Also, please clarify the pros and cons of using it.

**Comment by davelab6** [comment]
I agree with Felipe's point in the other issue that the syntax is secondary,and the primary thing is having good content. 

My reason for discussing simplification of syntax is to keep the content in one place, ie one string object, in order to make it easier to (a) edit and improve the content and (b) generate pretty docs from those DIY or with standard `__doc__` generating tools like Sphinx

**Comment by davelab6** [comment]
However a simple py string is totally unstructured, so a front matter or regex'able pattern could add structure

**Comment by davelab6** [comment]
By regex i mean that a link like `[Rationale](http://link.to.issue)` can be extracted if needed.

**Comment by graphicore** [comment]
docstring + front matter is not a big thing to implement¹. I think the pros are that there's a single place to put documentation information targeted at humans. Others are that we can treat the information centrally. I.e. the indentation from the `checkfunc.__doc__` is already removed, while `rationale="""` needs that treatment extra.
It would also remove the ambiguity between the `description`² part of a multi-line doc string and the `rationale` keyword (BTW: I never got why `rationale` was needed in spite of `description` , please elaborate). I believe the description section of the doc string is the right place to put the rationale, otherwise I wonder what to put there at all.
The only-docstring version is also very transparent on how the translation from doc-string to webpage will happen. The only negative point that I can see is that we can't put a finger on just the `rationale` section in the description text (without using regex or some other tricks) but since all in there is meant to be read by humans, I am convinced that proper, hierarchical structuring is all we need. For the yaml front-matter, we can figure out special handling for well known (and documented) keys. 
Making the front matter data accessible via a property of the check object is also no big deal.


## Examples

Here's a existing check with rationale and docsting-description. I believe the markdown structures the information better while keeping it together in one place.

```py
@check(
    id = 'com.google.fonts/check/153'
  , rationale = """
      Visually QAing thousands of glyphs by hand is tiring. Most glyphs can only
      be constructured in a handful of ways. This means a glyph's contour count
      will only differ slightly amongst different fonts, e.g a 'g' could either
      be 2 or 3 contours, depending on whether its double story or single story.
      However, a quotedbl should have 2 contours, unless the font belongs to a
      display family.
    """
)
def com_google_fonts_check_153(ttFont):
  """Check if each glyph has the recommended amount of contours.

  This check is useful to assure glyphs aren't incorrectly constructed.

  The desired_glyph_data module contains the 'recommended' countour count
  for encoded glyphs. The contour counts are derived from fonts which were
  chosen for their quality and unique design decisions for particular glyphs.

  In the future, additional glyph data can be included. A good addition would
  be the 'recommended' anchor counts for each glyph.
  """
```

docstring version, with some markdown formatting.

```py
@check(id = 'com.google.fonts/check/153')
def com_google_fonts_check_153(ttFont):
  """Check if each glyph has the recommended amount of contours.

  ## Rationale

  Visually QAing thousands of glyphs by hand is tiring. Most glyphs can only
  be constructured in a handful of ways. This means a glyph's contour count
  will only differ slightly amongst different fonts, e.g a 'g' could either
  be 2 or 3 contours, depending on whether its double story or single story.
  However, a quotedbl should have 2 contours, unless the font belongs to a
  display family.

  ## Details

  This check is useful to assure glyphs aren't incorrectly constructed.

  The desired_glyph_data module contains the 'recommended' countour count
  for encoded glyphs. The contour counts are derived from fonts which were
  chosen for their quality and unique design decisions for particular glyphs.

  In the future, additional glyph data can be included. A good addition would
  be the 'recommended' anchor counts for each glyph.
  """
```

---

Here's an existing check with more metadata:

```py
@check(
  id = 'com.google.fonts/check/163',
  rationale = """
    According to a Glyphs tutorial (available at
    https://glyphsapp.com/tutorials/multiple-masters-part-3-setting-up-instances),
    in order to make sure all versions of Windows recognize it as a valid
    font file, we must make sure that the concatenated length of the
    familyname (NAMEID_FONT_FAMILY_NAME) and style
    (NAMEID_FONT_SUBFAMILY_NAME)
    strings in the name table do not exceed 20 characters.
    """,
  misc_metadata = {
    # Somebody with access to Windows should make some experiments
    # and confirm that this is really the case.
    'affects': [('Windows', 'unspecified')],
    'request': 'https://github.com/googlefonts/fontbakery/issues/1488',
  }
)
  """Combined length of family and style must not exceed 20 characters."""
```


docstring version, we also use the front-matter proposal:

```py

@check(id='com.google.fonts/check/163')
def com_google_fonts_check_163(ttFont):
  """Combined length of family and style must not exceed 20 characters.

  ---
  # Somebody with access to Windows should make some experiments
  # and confirm that this is really the case.
  affects: 
    - ['Windows', 'unspecified']
  request: https://github.com/googlefonts/fontbakery/issues/1488
  ---

  ## Rationale

  According to a Glyphs tutorial (available at
  https://glyphsapp.com/tutorials/multiple-masters-part-3-setting-up-instances),
  in order to make sure all versions of Windows recognize it as a valid
  font file, we must make sure that the concatenated length of the
  familyname (NAMEID_FONT_FAMILY_NAME) and style
  (NAMEID_FONT_SUBFAMILY_NAME)
  strings in the name table do not exceed 20 characters.
"""

```

## Footnotes



¹ implement front matter parsing
```py
def extractFrontmatter(description):
    default =  description, {}
    lines = description.splitlines(True)
    if lines[0].strip() != '---':
        return default

    lines = lines[1:]
    yamlData=None
    for i, line in enumerate(lines):
        if line.strip() == '---':
            yamlData = ''.join(lines[:i])
            lines = lines[i+1:]
            break
    if yamlData is None:
        return default

    try:
        parsed = yaml.load(yamlData)
    except yaml.YAMLError as exc:
        # Maybe log exc here.
        return default

    return ''.join(lines), parsed
```

² [pep 257](https://www.python.org/dev/peps/pep-0257/#id17): "Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description."

**Comment by graphicore** [comment]
From @felipesanches in https://github.com/googlefonts/fontbakery/issues/1584#issuecomment-385291162

> I looked up this "front-matter yaml" thing this afternoon but we already have a good solution for our needs. All the data is already available on runtime and now better organized with the misc_metadata field suggested by Lasse and adopted on the latest commit above.
>
>We should focus efforts on the real important thing which is collecting more metadata such as rationales for the checks rather than the technology adopted or syntactic sugar

my five cents:

> is already available on runtime

Not true. unknown keywords are going into the void. Especially `rationale` and `misc_metadata` are not available.

> better organized with the misc_metadata field suggested by Lasse

yeah, I suggested this, but I think front matter is an adequate alternative.

> We should focus efforts on the real important thing which is collecting more metadata

I agree that this is bikeshedding, but this is also an  anachronism. Why aren't we using the docstrings in the first place, they were there before the `rationale` keyword etc. appeared. I don't remember a proposal, examples and pro-and cons for these. It's a bit unfair to create precedents and then act like they are the status quo. So what are the pros on these constructor keyword arguments? We never heard about these.

> rather than the technology adopted or syntactic sugar

See my answer above, it's no big thing to do, the *efforts* are no big deal. In fact, writing all these bike shedding commentary took probably already longer than implementing it. 

In the end, I would hope that collecting metadata is easier when done in docstings only, as one mardown formatted article, structured hierarchically.

**Comment by graphicore** [comment]
Relabeled as `P3 soon` as I'm planning to use this kind of data soon.

**Comment by simoncozens** [comment]
We instead went the other way and had more structured information in the check decorator.
